docs/writing_plugins.md

```diff
@@ -0,0 +1,550 @@
+Writing Plugins
+========================
+
+## Introduction
+
+This package helps to create your own code generator plugin.  At a high level, the process is:
+
+1.  Implement a TypeScript generator function, `generateTs`, which is used to generate TypeScript files
+2.  Pass this function along with some static configuration values (and optional functions) to `createEcmaScriptPlugin` from the `@bufbuild/protoplugin` package.
+3.  Create an executable file which runs your plugin.
+4.  Specify this in your `package.json` file.
```

Comment:
This is a great summary, but can we give an introduction to plugins first? For example:

> Code generator plugins are a unique feature of protocol buffer compilers like protoc and the buf CLI. With a plugin, you can generate RPC clients and server stubs, mappings from protobuf to SQL, validation code, and pretty much anything else you can think of.

We can also quickly summarize how plugins work:
> The contract between the protobuf compiler and a code generator plugin is defined in plugin.proto. Plugins are simple executables (typically on your $PATH) that are named `protoc-gen-x`, where `x` is the name of the language or feature that the plugin provides. The protobuf compiler parses the protobuf files, and invokes the plugin, sending a `CodeGeneratorRequest` on standard in, and expecting a `CodeGeneratorResponse` on standard out. The request contains a set of descriptors (see descriptor.proto) - an abstract version of the parsed protobuf files. The response contains a list of files, each having a name and text content.

This lets us explain what @bufbuild/protoplugin adds on top:
- A simplified view of the descriptors that resolves references, hides implementation details like synthetic map entries, and provides simple access to comments.
- Automatic import and export management for ECMAScript.
- First class support for referencing code generated by @bufbuild/protoc-gen-es.
- Automatic transpilation of TypeScript to JavaScript and TypeScript declaration files.

(Not necessarily in the form of this bullet list - it's fine if our documentation makes these points obvious)